New blog: what is continuous profiling? (#60)
pyroscope-io/docs: blog/what-is-continuous-profiling.mdx

---
title: What is continuous profiling?
description: An introductory guide to continuous profiling
sidebar_label: What is continuous profiling
image: https://user-images.githubusercontent.com/23323466/143324845-16ff72df-231e-412d-bd0a-38ef2e09cba8.gif

> slug: /intro-to-continuous-profiling

Reviewer: let's make the slug match the blog title:
```suggestion
slug: /what-is-continuous-profiling
```

date: "2022-04-26"

authors:
  - name: Uchechukwu obasi
    url: https://github.com/thisisobate
    image_url: https://avatars.githubusercontent.com/u/29557702?v=4

tags: [observability, flamegraphs, profiling, go]
---

Coming from a background working as a frontend developer at Grafana I'm no stranger to open source performance monitoring. I was part of a team that was responsible for the overall user experience of Grafana and performance was one of the key considerations. Along the line, I learned about a debugging technique known as profiling for monitoring application performance and fell in love ever since.

![chrome browser profiler](https://user-images.githubusercontent.com/29557702/165289655-58cd9a0b-8848-4e03-b7e5-afdf822cf577.png)

> ### Introduction

Reviewer: Even though it's slightly repetitive I think its better and more descriptive to make this title indicate the content that will follow:

```suggestion
### What is continuous profiling
```

As a frontend developer, my experience with profiling was limited to the browser. However, in the course of my study, I discovered a new pattern of profiling that seems exciting– continuous profiling. Similar to how you use the profiler in the dev console to understand frontend performance issues, continuous profiling allows you to profile servers from various languages 24/7 and be able to understand resource usage at any particular time. 

While continuous profiling is new to many, the concept is actually relatively old. In 2010, Google released a paper titled [“Google Wide profiling: A continuous profiling infrastructure for data centers”](https://research.google/pubs/pub36575/) where they make the case for the value of adding continuous profiling to your applications. 

### Industry traction for continuous profiling

> Since then, many major performance monitoring solutions have joined them in releasing continuous profiling products. At the same time, many other companies and VCs are making major investments in Continuous Profiling as well to catch up to the popular demand for this tool.

Reviewer: ```suggestion
Since then, many major performance monitoring solutions have joined them in releasing continuous profiling products. As time has gone on, the continuous profiling space has been getting increasingly popular as various companies/VCs are more frequently making major investments in Continuous Profiling to keep up with demand for this type of monitoring.
```

It’s not just major cloud providers and venture capitalists who are excited about continuous profiling, our industry leaders are super pumped about continuous profiling too! 

<Tweet tweetLink="cra/status/1474083458943266817" />

Chris Aniszczyk, CTO of the Cloud Native Computing Foundation (CNCF) agrees that continuous profiling is an integral part of the observability stack.

<Tweet tweetLink="mhausenblas/status/1462047847776407556" />

This poll by Michael Hausenblas, solution engineering lead at AWS is good evidence that continuous performance profiling is becoming a culture for most engineering teams.

<Tweet tweetLink="CarlBergquist/status/1490351369890504710" hideConversation={true}/>

Even Carl Bergquist, Principal Engineer at Grafana, predicts that the popular FOSDEM conference will see more profiling talks moving forward.

<Tweet tweetLink="lizthegrey/status/1449115630213341186" hideConversation={true} />

And like Liz Fong-Jones, Principal developer advocate at Honeycomb, many in the tech world are “buzzing” about continuous profiling!

**Simply put, continuous profiling is the future of application performance assessment**

### Continuous profiling using Pyroscope

> At Pyroscope, we actually [profile our own servers using Pyroscope](https://pyroscope.io/blog/profile-continuous-profiler/) and have found many cases where profiling has saved us a lot of time and money by identifying performance issues. Pyroscope allows: 

Reviewer: Did you try something like `/blog/profile-continuous-profiler` I think thats the proper way to do this in docasaurus (maybe double check that it works though)

```suggestion
At Pyroscope, we actually [profile our own servers using Pyroscope](/blog/profile-continuous-profiler) and have found many cases where profiling has saved us a lot of time and money by identifying performance issues. Pyroscope allows: 
```

- Developers
  - Understand the call tree of their application
  - Locate and fix memory leaks
  - Resolve issues with high CPU utilization
- Engineering Managers
  - Easily spot performance regression in their codebase
  - 360 performance coverage of their infrastructure
  - Create a performance-driven culture where performance testing is carried out throughout the software development cycle.
- Executives
  - Have access to performance insight for better decision making
  - Reduce cost allocation for computing resources
  - Get real-time analysis of their compute resource utilization

Pyroscope also provides you with multi-language support, ad-hoc profiling, profile sharing, and more!

![Pyroscope GIF Demo](https://user-images.githubusercontent.com/23323466/143324845-16ff72df-231e-412d-bd0a-38ef2e09cba8.gif)

### What makes Pyroscope unique?

> Pyroscope is open-source which means that it’s easier to get started and try out than cloud providers profilers. Pyroscope’s custom storage engine is language-agnostic so profilers such as [rbspy](https://rbspy.github.io/), [eBPF](https://ebpf.io/), and even Bloomberg's new memory profiler [Memray](https://bloomberg.github.io/memray/), can easily send data to Pyroscope and have it stored and queried efficiently. 

Reviewer: ```suggestion
Pyroscope is open-source which means that it’s easier to get started and try out than cloud providers profilers. Pyroscope’s [custom storage engine](docs/storage-design/) is language-agnostic so _any profiling data_ ranging from [rbspy](https://rbspy.github.io/) to [eBPF](https://ebpf.io/) to even new profilers such as Bloomberg's [Memray](https://bloomberg.github.io/memray/) data can easily send data to Pyroscope and have it compressed, stored and queried efficiently. 
```

---
title: What is continuous profiling?
description: A beginner introductory guide in profiling
sidebar_label: What is continuous profiling
image: https://user-images.githubusercontent.com/23323466/143324845-16ff72df-231e-412d-bd0a-38ef2e09cba8.gif
slug: /intro-to-continuous-profiling
date: "2022-04-26"

authors:
  - name: Uchechukwu obasi
    url: https://github.com/thisisobate
    image_url: https://avatars.githubusercontent.com/u/29557702?v=4

tags: [observability, flamegraphs, profiling, go]
---

> ![browser-profiler](https://user-images.githubusercontent.com/29557702/165289655-58cd9a0b-8848-4e03-b7e5-afdf822cf577.png)

Author: ok

### Introduction
“Profiling” is a dynamic method of analyzing the complexity of a program, such as CPU utilization or the frequency and duration of function calls. With profiling, you can locate exactly which parts of your application are consuming the most resources. “Continuous profiling” is a more powerful version of profiling that adds the dimension of time. By understanding your system's resources over time, you can then locate, debug, and fix issues related to performance. 

As a frontend developer, my experience with profiling was limited to the browser. However, in the course of my study, I discovered a new pattern of profiling that seems exciting– continuous profiling. Similar to how you use the profiler in the dev console to understand frontend performance issues, continuous profiling allows you to profile servers from various languages 24/7 and be able to understand resource usage at any particular time. 

> ![gwp-diagram-edited](https://user-images.githubusercontent.com/29557702/165291203-68d7acff-4b05-41f9-9d24-9ea4bb4fa48d.png)

Reviewer: Lets remove this image I think it is out of style with our other images and doesn't add much

![cp_timeline_08-01](https://user-images.githubusercontent.com/29557702/165291672-88805173-4bf4-4029-9b6c-3b31951733bd.png)

It’s not just major cloud providers and venture capitalists who are excited about continuous profiling, our industry leaders are super pumped about continuous profiling too! 

<Tweet tweetLink="cra/status/1474083458943266817" />

Chris Aniszczyk, CTO of the Cloud Native Computing Foundation (CNCF) agrees that continuous profiling is an integral part of the observability stack.

<Tweet tweetLink="mhausenblas/status/1462047847776407556" />

This poll by Michael Hausenblas, solution engineering lead at AWS is good evidence that continuous performance profiling is becoming a culture for most engineering teams.

<Tweet tweetLink="CarlBergquist/status/1490351369890504710" hideConversation={true}/>

Even Carl Bergquist, Principal Engineer at Grafana, predicts that the popular FOSDEM conference will see more profiling talks moving forward.

<Tweet tweetLink="lizthegrey/status/1449115630213341186" hideConversation={true} />

And like Liz Fong-Jones, Principal developer advocate at Honeycomb, many in the tech world are “buzzing” about continuous profiling!

> ## Simply put, continuous profiling is the future of application performance assessment

Reviewer: Let's just make this bold as opposed to making it a header
```suggestion
**Simply put, continuous profiling is the future of application performance assessment**
```

### Continuous profiling using Pyroscope

> At Pyroscope, we actually profile our own servers using Pyroscope and have found many cases where profiling has saved us a lot of time and money by identifying performance issues. Pyroscope allows: 

Reviewer: Add link to https://pyroscope.io/blog/profile-continuous-profiler/

- Developers
  - Understand the call tree of their application
  - Locate and fix memory leaks
  - Resolve issues with high CPU utilization
- Engineering Managers
  - Easily spot performance regression in their codebase
  - 360 performance coverage of their infrastructure
  - Create a performance-driven culture where performance testing is carried out throughout the software development cycle.
- Executives
  - Have access to performance insight for better decision making
  - Reduce cost allocation for computing resources
  - Get real-time analysis of their compute resource utilization

Pyroscope also provides you with multi-language support, ad-hoc profiling, profile sharing, and more!

![Pyroscope GIF Demo](https://user-images.githubusercontent.com/23323466/143324845-16ff72df-231e-412d-bd0a-38ef2e09cba8.gif)

### What makes Pyroscope unique?

> Pyroscope is open-source which means that it’s easier to get started and try out than cloud providers profilers. Pyroscope’s custom storage engine is language-agnostic so profilers such as Rbspy, eBPF, and even Bloomberg's new memory profiler Memray, can easily send data to Pyroscope and have it stored and queried efficiently. Pyroscope is also built to be easy to use by everyone. The pyroscope UI in particular was designed to be super intuitive for anyone to play around, analyze, and understand profiles.

Reviewer: lets add links for rbspy (also make it all lowercase) and memray

> Pyroscope is open-source which means that it’s easier to get started and try out than cloud providers profilers. Pyroscope’s custom storage engine is language-agnostic so profilers such as Rbspy, eBPF, and even Bloomberg's new memory profiler Memray, can easily send data to Pyroscope and have it stored and queried efficiently. Pyroscope is also built to be easy to use by everyone. The pyroscope UI in particular was designed to be super intuitive for anyone to play around, analyze, and understand profiles.

Reviewer: lets add links for rbspy (also make it all lowercase) and memray also lets add a line break
```suggestion
Pyroscope is open-source which means that it’s easier to get started and try out than cloud providers profilers. Pyroscope’s custom storage engine is language-agnostic so profilers such as Rbspy, eBPF, and even Bloomberg's new memory profiler Memray, can easily send data to Pyroscope and have it stored and queried efficiently. 

Pyroscope is also built to be easy to use by everyone. The pyroscope UI in particular was designed to be super intuitive for anyone to play around, analyze, and understand profiles.
```

Whether you are operating at a small scale or enterprise level, the power of continuous profiling can quickly and easily reduce infrastructure costs and improve scalability.  With Pyroscope, you can build high-performance applications that increase customer satisfaction and put value back into your business. 

### How to get started with Pyroscope

> Given my experience with profiling in the browser, the easiest way for me to get started with continuous profiling was to install pyroscope. To learn how to perform continuous profiling in your code or application, I would suggest you run the [examples in the examples folder on GitHub](https://github.com/pyroscope-io/pyroscope/tree/main/examples). 

Reviewer 2: > Given my experience with profiling in the browser, the easiest way for me to get started with continuous profiling was to install pyroscope.

I like the personal touch, but I don't really get the connection between experience with profiling in the browser and it helping when installing pyroscope.

Author: Basically, I was just trying to say that regardless of my limited experience with profilers, I was still able to do continuous profiling and Pyrsocope made it so much easy for me.

> Given my experience with profiling in the browser, the easiest way for me to get started with continuous profiling was to install pyroscope. To learn how to perform continuous profiling in your code or application, I would suggest you run the [examples in the examples folder on GitHub](https://github.com/pyroscope-io/pyroscope/tree/main/examples). 

Reviewer: ```suggestion
Coming in as being relatively new to profiling myself, the easiest way for me to get started with continuous profiling was to install pyroscope. To learn how to use continuous profiling in your code or application, I would suggest you run the relevant dockerized [examples from the examples folder on GitHub](https://github.com/pyroscope-io/pyroscope/tree/main/examples). 
```

> Pyroscope has offered me useful features that make profiling fun and exciting. With Pyroscope, I’ve been able to understand how my application, server, and code are running. The best part is that I can use Pyroscope without paying any money; it’s completely free!

Reviewer: Add another line break here
```suggestion

Pyroscope has offered me useful features that make profiling fun and exciting. With Pyroscope, I’ve been able to understand how my application, server, and code are running. The best part is that I can use Pyroscope without paying any money; it’s completely free!
```

> To get started with Pyroscope, take a look at our installation guide [here](https://pyroscope.io/docs/installing-pyroscope-overview/)

Reviewer: Removed this part to move it above
```suggestion
```

Author: you mean I should move it just right below the "...get started with pyroscope" heading?

Author: Was considering removing this section entirely

Author: anyways...nvm 😅

> Follow us on [Github](https://github.com/pyroscope-io/pyroscope) or [Twitter](https://twitter.com/PyroscopeIO) or join our [slack](https://pyroscope.io/slack) for updates on this!

Reviewer: ```suggestion
If you need any help getting started join the [community slack](https://pyroscope.io/slack) or follow Pyroscope on [Github](https://github.com/pyroscope-io/pyroscope) and [Twitter](https://twitter.com/PyroscopeIO)!
```
